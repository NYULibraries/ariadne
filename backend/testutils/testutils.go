package testutils

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

//go:embed testdata/test-cases.json
var TestCasesJSON []byte

type TestCase struct {
	// Identifier used for fixture and golden file basename.
	Key string
	// Human-readable name/description of test case
	Name string
	// OpenURL querystring
	QueryString string
}

var TestCases []TestCase

var testutilsPath string

// We need to get the absolute path to this package in order to enable the function
// for golden file and fixture file retrieval to be called from other packages
// which would not be able to resolve the hardcoded relative paths used here.
func init() {
	// The `filename` string is the absolute path to this source file, which should
	// be located at the root of the package directory.
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		panic("ERROR: `runtime.Caller(0)` failed")
	}

	// Get the path to the parent directory of this file.  Again, this is assuming
	// that this `init()` function is defined in a package top level file -- or
	// more precisely, that this file is in the same directory at the `testdata/`
	// directory that is referenced in the relative paths used in the functions
	// defined in this file.
	testutilsPath = filepath.Dir(filename)

	err := json.Unmarshal(TestCasesJSON, &TestCases)
	if err != nil {
		panic(fmt.Sprintf("Error reading test cases file: %s", err))
	}

}

func GetGoldenValue(testCase TestCase) (string, error) {
	return GetTestdataFileContents(GoldenFile(testCase))
}

func GetPrimoFakeResponse(testCase TestCase) (string, error) {
	return GetTestdataFileContents(primoFakeResponseFile(testCase))
}

func GetSFXFakeResponse(testCase TestCase) (string, error) {
	return GetTestdataFileContents(sfxFakeResponseFile(testCase))
}

func GetTestdataFileContents(filename string) (string, error) {
	bytes, err := os.ReadFile(filename)

	if err != nil {
		return filename, err
	}

	return string(bytes), nil
}

func GoldenFile(testCase TestCase) string {
	return testutilsPath + "/testdata/golden/" + testCase.Key + ".json"
}

func primoFakeResponseFile(testCase TestCase) string {
	return testutilsPath + "/testdata/fixtures/primo-fake-responses/" + testCase.Key + ".xml"
}

func sfxFakeResponseFile(testCase TestCase) string {
	return testutilsPath + "/testdata/fixtures/sfx-fake-responses/" + testCase.Key + ".xml"
}
