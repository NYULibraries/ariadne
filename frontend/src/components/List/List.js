import 'bootstrap/dist/css/bootstrap.min.css';
import { useEffect } from 'react';
import { getCoverageStatement } from '../../aux/helpers';
import useApi from '../../hooks/useApi';
import linksApi from '../../api/fetchData';

const List = () => {
  const backendClient = useApi(linksApi.fetchData);

  useEffect(() => {
    backendClient.fetchResource();
    // We want to fetch the resource only once when component mounts, so we pass an empty array as a second argument
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <div className="jumbotron">
        <div className="container text-center">
          <p>Displaying search results...</p>
          <p>Note: Alternate titles might have matched your search terms</p>
        </div>
      </div>
      <div className="i-am-centered">
        <div className="list-group">
          {/* TODO: we could put a spinner here: */}
          {backendClient.loading && <div>Loading...</div>}
          {backendClient.error && <div className="i-am-centered">{backendClient.error}</div>}
          {backendClient.resource?.map((link, idx) => (
            <div key={idx} className="list-group-item list-group-item-action flex-column" style={{ border: 'none' }}>
              <div className="row">
                <h6>
                  <a href={link.target_url} target="_blank" rel="noopener noreferrer">
                    {link.target_public_name}
                  </a>
                </h6>
                <small>{getCoverageStatement(link)}</small>
              </div>
            </div>
          ))}
          {backendClient.resource?.length === 0 && <div className="i-am-centered">No results found</div>}
        </div>
        {backendClient.resourceLastElement && (
          <div className="ask-librarian">
            <h6>
              <a href={backendClient.resourceLastElement.target_url} target="_blank" rel="noopener noreferrer">
                {backendClient.resourceLastElement.target_public_name}
              </a>
            </h6>
          </div>
        )}
      </div>
    </>
  );
};

export default List;
