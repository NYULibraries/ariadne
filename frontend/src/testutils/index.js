const fs = require('fs');
const path = require('path');

// Prefer actual backend/api/testdata/ files if they exist (i.e. running tests on
// local machine while working in the repo).  If in a container, the backend/api/testdata/
// files will have been copied into this directory.
const BACKEND_API_TESTDATA_DIR_IN_REPO = path.join(
    __dirname,
    '..',
    '..',
    '..',
    'backend',
    'api',
    'testdata',
    'server',
);
const BACKEND_API_TESTDATA_DIR_IN_CONTAINER = path.join(
    __dirname,
    'backend-testdata',
);
const BACKEND_API_TESTDATA_DIR = fs.existsSync(BACKEND_API_TESTDATA_DIR_IN_REPO) ?
                                 BACKEND_API_TESTDATA_DIR_IN_REPO :
                                 BACKEND_API_TESTDATA_DIR_IN_CONTAINER;

const BACKEND_API_TEST_CASES_GOLDEN_FILES_DIR = path.join(BACKEND_API_TESTDATA_DIR, 'golden');
const BACKEND_API_TEST_CASES_INDEX = path.join(BACKEND_API_TESTDATA_DIR, 'test-cases.json');

function getTestCases() {
    const testCasesFromAriadneBackendApi = require(BACKEND_API_TEST_CASES_INDEX);

    testCasesFromAriadneBackendApi.forEach( testCase => {
        testCase.response = getResponse(testCase.key);
    });

    return testCasesFromAriadneBackendApi;
}

function getResponse(key) {
    const goldenFile = path.join(BACKEND_API_TEST_CASES_GOLDEN_FILES_DIR, `${key}.json`);

    return require(goldenFile);
}

export {
    getTestCases,
};
