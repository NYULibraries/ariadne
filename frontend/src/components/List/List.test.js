import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import '@testing-library/jest-dom';
import List from './List';
import { getTestCases } from '../../testutils';

const testCases = getTestCases();
testCases.forEach( testCase => {
  describe(testCase.name, () => {

    beforeEach(() => {
      delete window.location;
      window.location = new URL(`${process.env.REACT_APP_API_URL}?${testCase.queryString}`);
    });

    afterEach(() => {
      delete window.location;
      window.location = new URL('http://localhost:3000');
    });

    test('renders with a className of list-group', () => {
      const { container } = render(<List />);
      expect(container.getElementsByClassName('list-group').length).toBe(1);
    });

    test.skip('renders correctly', async () => {
      const actual = render(<List />);
      await waitForElementToBeRemoved(() => screen.getByText(/Loading.../i));
      expect(actual.asFragment()).toMatchSnapshot();
    });

    test('renders the search results', async () => {
      render(<List />);
      const linkElement = await waitFor(() => screen.getByText(/Displaying search results.../i));
      expect(linkElement).toBeInTheDocument();
    });

    test('renders Loading...', () => {
      render(<List />);
      const linkElement = screen.getByText(/Loading/i);
      expect(linkElement).toBeInTheDocument();
    });

    test('renders Ask a Librarian', async () => {
      render(<List />);
      const linkElement = await waitFor(() => screen.getByText(/Ask a Librarian/i));
      expect(linkElement).toBeInTheDocument();
    });

    test('Loading... no longer present in the DOM after loading data', async () => {
      const { getByText } = render(<List />);
      await waitForElementToBeRemoved(() => getByText(/Loading/i));
    });

  });
});
