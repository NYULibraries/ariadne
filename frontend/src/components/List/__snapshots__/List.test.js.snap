// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Corriere Fiorentino renders correctly 1`] = `
<DocumentFragment>
  <div
    class="jumbotron"
  >
    <div
      class="container text-center"
    >
      <p>
        Displaying search results...
      </p>
      <p>
        Note: Alternate titles might have matched your search terms
      </p>
    </div>
  </div>
  <div
    class="i-am-centered"
  >
    <div
      class="list-group"
    >
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="
                    http://proxy.library.nyu.edu/login?url=https://www.pressreader.com/italy/corriere-fiorentino
                "
              rel="noopener noreferrer"
              target="_blank"
            >
              PressReader
            </a>
          </h6>
          <small />
        </div>
      </div>
    </div>
    <div
      class="ask-librarian"
    >
      <h6>
        <a
          href="http://library.nyu.edu/ask/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Ask a Librarian
        </a>
      </h6>
    </div>
  </div>
</DocumentFragment>
`;

exports[`The New Yorker renders correctly 1`] = `
<DocumentFragment>
  <div
    class="jumbotron"
  >
    <div
      class="container text-center"
    >
      <p>
        Displaying search results...
      </p>
      <p>
        Note: Alternate titles might have matched your search terms
      </p>
    </div>
  </div>
  <div
    class="i-am-centered"
  >
    <div
      class="list-group"
    >
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=http://archives.newyorker.com/#folio=C1"
              rel="noopener noreferrer"
              target="_blank"
            >
              E Journal Full Text
            </a>
          </h6>
          <small>
            Available from 1925
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=http://gateway.proquest.com/openurl?rft_id=41130&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&url_ver=Z39.88-2004&res_dat=xri%3Apqm&rfr_id=info%3Axri%2Fsid%3Aprimo&genre=journal"
              rel="noopener noreferrer"
              target="_blank"
            >
              Art, Design & Architecture Collection
            </a>
          </h6>
          <small>
            Available from 2002/11/04
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=http://find.galegroup.com/openurl/openurl?url_ver=Z39.88-2004&url_ctx_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Actx&rft.issn=0028-792X&res_id=info%3Asid%2Fgale%3ASHAX&req_dat=info%3Asid%2Fgale%3Augnid%3Anew64731&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&ctx_enc=info%3Aofi%3Aenc%3AUTF-8&rft.date=2002&rft.jtitle=New+Yorker"
              rel="noopener noreferrer"
              target="_blank"
            >
              Gale Cengage Shakespeare Collection Periodicals
            </a>
          </h6>
          <small>
            Available from 2002
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://link.gale.com/apps/pub/1161/ITOF?u=nysl_me_newyorku"
              rel="noopener noreferrer"
              target="_blank"
            >
              Gale General OneFile
            </a>
          </h6>
          <small>
            Available from 2002/01/14
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://link.gale.com/apps/pub/1161/LitRC?u=new64731"
              rel="noopener noreferrer"
              target="_blank"
            >
              Gale Literature Resource Center
            </a>
          </h6>
          <small>
            Available from 1978/01/01  until 1978/12/31. Available from 1982/01/01  until 1982/12/31. Available from 1989/01/01  until 1989/12/31. Available from 1996/01/01  until 1996/12/31. Available from 2002/01/01
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://advance.lexis.com/api/search/advanced?source=MTA2OTUwNg&identityprofileid=W4HVBF32601"
              rel="noopener noreferrer"
              target="_blank"
            >
              Lexis Advance US
            </a>
          </h6>
          <small>
            Available from 1999
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=http://archives.newyorker.com/#folio=C1"
              rel="noopener noreferrer"
              target="_blank"
            >
              Miscellaneous Ejournals
            </a>
          </h6>
          <small>
            Available from 1925
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://gateway.proquest.com/openurl?rft_id=41130&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&url_ver=Z39.88-2004&res_dat=xri%3Apqm&rfr_id=info%3Axri%2Fsid%3Aprimo&genre=journal"
              rel="noopener noreferrer"
              target="_blank"
            >
              Music & Performing Arts Collection
            </a>
          </h6>
          <small>
            Available from 2002/11/04
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://gateway.proquest.com/openurl?rft_id=16493&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&url_ver=Z39.88-2004&rfr_id=info%3Axri%2Fsid%3Aprimo&res_dat=xri%3Apqm&genre=journal"
              rel="noopener noreferrer"
              target="_blank"
            >
              Music & Performing Arts Collection
            </a>
          </h6>
          <small>
            Available from 2001/08/20  until 2017/01/02
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=http://www.newyorker.com/archive"
              rel="noopener noreferrer"
              target="_blank"
            >
              OpinionArchives
            </a>
          </h6>
          <small>
            Available from 1925
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://gateway.proquest.com/openurl?rft_id=41130&rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&res_dat=xri%3Apqm&rfr_id=info%3Axri%2Fsid%3Aprimo&url_ver=Z39.88-2004&genre=journal"
              rel="noopener noreferrer"
              target="_blank"
            >
              ProQuest Central
            </a>
          </h6>
          <small>
            Available from 2002/11/04
          </small>
        </div>
      </div>
      <div
        class="list-group-item list-group-item-action flex-column"
        id="border-style"
      >
        <div
          class="row"
        >
          <h6>
            <a
              href="http://proxy.library.nyu.edu/login?url=https://global.factiva.com/en/du/headlines.asp?XSID=S001dbr5DEs5DEmN9MpMD6mNDVyMHmnRsIuMcNG1pRRQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQQAA&CurrentSourcesDesc=sc_u_gtny%2CNew+Yorker&CurrentSources=U%7Cgtny"
              rel="noopener noreferrer"
              target="_blank"
            >
              Factiva
            </a>
          </h6>
          <small>
            Available from 1997
          </small>
        </div>
      </div>
    </div>
    <div
      class="ask-librarian"
    >
      <h6>
        <a
          href="http://library.nyu.edu/ask/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Ask a Librarian
        </a>
      </h6>
    </div>
  </div>
</DocumentFragment>
`;
